app/api/text-to-sql/agent.ts

```diff
@@ -0,0 +1,84 @@
+import { Agent } from "@mastra/core/agent";
+import { threadMemory } from "../memory";
+import { flash } from "../../utils/models";
+import { createTool } from "@mastra/core";
+import { z } from "zod";
+import { databaseSchema, getDatabaseSchema } from "./shared";
+import { db } from "../../db/db";
+import { sql } from "drizzle-orm";
+import { pluralize } from "../../utils/string";
+
+const schemaFetcher = createTool({
+  id: "schema-fetcher",
+  description: "Fetch the schema of the database",
+  inputSchema: z.object({
+    // TODO: Add database name/connection string
+  }),
+  outputSchema: z.object({
+    schema: databaseSchema,
+  }),
+  execute: async () => {
+    const schema = await getDatabaseSchema();
+    return {
+      schema: schema,
+    };
+  },
+});
+
+const executeQuery = createTool({
+  id: "execute-query",
+  description: "Execute a SQL query",
+  inputSchema: z.object({
+    query: z.string(),
+  }),
+  outputSchema: z.object({
+    results: z.array(z.any()),
+  }),
+  execute: async ({ context }) => {
+    // Validate the query
+    const bannedKeywords = ["DROP", "DELETE", "UPDATE", "ALTER", "TRUNCATE"];
```

Comment:
Riiiight - flying into the danger zone with text to sql agents

Reply:
Absolutely, probably could add more too, and you'd want to have a dedicated Postgres user too with the strictest permissions needed so that even if somehow they get around this it's all under control!